lib/install.js

```diff
@@ -897,28 +902,34 @@ function clean() {
 }
 exports.clean = clean;
 
+/* Like concat, but in place */
+function into(extras) {
+  var list = this;
+  extras.forEach(function(item) { list.push(item); });
+}
+
 // depthCheck returns true to keep going (files being ignored), false to add the dir to the flat list
 function readDirWithDepth(dir, depthCheck) {
   var flatDirs = [];
-  return asp(fs.readdir)(dir)
-  .catch(function(e) {
-    if (e.code === 'ENOTDIR')
-      return;
-    throw e;
+  return asp(fs.lstat)(dir).then(function(fileInfo) {
+    if (fileInfo.isSymbolicLink())
+      return Promise.resolve([dir]);
+
+    if (!fileInfo.isDirectory())
+      return Promise.resolve([]);
+
+    return asp(fs.readdir)(dir);
   })
   .then(function(files) {
     if (!files)
       return [];
     return Promise.all(files.map(function(file) {
-      if (depthCheck(file))
-        return readDirWithDepth(path.resolve(dir, file), depthCheck)
-        .then(function(dirs) {
-          dirs.forEach(function(subDir) {
-            flatDirs.push(subDir);
-          });
-        });
-      else
-        flatDirs.push(path.resolve(dir, file));
+      var filepath = path.resolve(dir, file);
+      if (dir != file && depthCheck(filepath)) {
+        return readDirWithDepth(filepath, depthCheck).then(into.bind(flatDirs));
```

Comment:
I really prefer keeping functions inline unless they have to be shared for readability.